File: src/zombie/subject_contents/procedures/parsers.py
"""Parsers for different types of timeline events."""

from datetime import timedelta
import pandas as pd

from zombie.subject_contents.procedures.fiber_implant_parser import has_fiber_implants


NO_END_DATE_DURATION = timedelta(days=1)  # Default duration for events without end date


def normalize_timestamp(ts):
    """Convert timestamp to timezone-aware UTC for consistent comparisons."""
    dt = pd.to_datetime(ts)
    # If timezone-naive, assume UTC
    if dt.tz is None:
        return dt.tz_localize("UTC")
    # If timezone-aware, convert to UTC
    return dt.tz_convert("UTC")


def parse_birth(subject):
    """Parse birth event from subject data."""
    if not subject:
        return None

    subject_details = subject.get("subject_details", {})
    if not subject_details:
        return None

    dob = subject_details.get("date_of_birth")
    if not dob:
        return None

    try:
        dob_date = normalize_timestamp(dob)
        # Birth is instantaneous - show as a wider line for visibility
        return {
            "start": dob_date,
            "end": dob_date + NO_END_DATE_DURATION,  # 2 days width for visibility
            "event": "Birth",
            "type": "Birth",
            "details": f"Subject {subject.get('subject_id', 'Unknown')} born",
            "data": subject_details,
        }
    except Exception as e:
        print(f"[parse_birth] Error parsing birth: {e}")
        return None


def parse_procedure(proc):
    """Parse procedure event from procedure data."""
    if not proc:
        return None

    proc_type = proc.get("object_type", "Procedure")
    start_date = proc.get("start_date")
    end_date = proc.get("end_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)

        # Use end_date if available, otherwise use a fixed width for visibility
        if end_date:
            end_dt = normalize_timestamp(end_date)
        else:
            # Use 2 days width for visibility/clickability, like birth events
            end_dt = start_dt + NO_END_DATE_DURATION

        # Get procedure details
        if proc_type == "Surgery":
            sub_procs = proc.get("procedures", [])
            if sub_procs:
                detail_parts = [sp.get("object_type", "Unknown") for sp in sub_procs if sp]
                detail_str = ", ".join(detail_parts) if detail_parts else "Surgery"
            else:
                detail_str = "Surgery"
        else:
            detail_str = proc_type

        return {
            "start": start_dt,
            "end": end_dt,
            "event": proc_type,
            "type": proc_type,
            "details": detail_str,
            "data": proc,
        }
    except Exception as e:
        print(f"[parse_procedure] Error parsing procedure: {e}")
        return None


def parse_session(data_desc):
    """Parse session event from data description."""
    from datetime import datetime

    if not data_desc:
        return None

    creation_time = data_desc.get("creation_time")
    if not creation_time:
        return None

    try:
        dt = datetime.fromisoformat(creation_time.replace("+00:00", ""))
        start_dt = pd.to_datetime(dt)
        # Sessions typically last hours
        end_dt = start_dt + timedelta(hours=4)

        return {
            "start": start_dt,
            "end": end_dt,
            "event": "Session",
            "type": "Session",
            "details": f"Data session: {data_desc.get('name', 'Unknown')}",
            "data": data_desc,
        }
    except Exception as e:
        print(f"[parse_session] Error parsing session: {e}")
        return None


def parse_acquisition(acquisition):
    """Parse acquisition event with start and end times."""
    from datetime import datetime

    if not acquisition:
        return None

    start_time = acquisition.get("acquisition_start_time")
    end_time = acquisition.get("acquisition_end_time")

    if not start_time or not end_time:
        return None

    try:
        start_dt = normalize_timestamp(start_time)
        end_dt = normalize_timestamp(end_time)

        # Get acquisition type or session type for the label
        acq_type = acquisition.get("acquisition_type") or acquisition.get("session_type", "Acquisition")
        protocol = acquisition.get("protocol_name", "")

        event_label = f"{acq_type}"
        if protocol:
            event_label += f" ({protocol})"

        return {
            "start": start_dt,
            "end": end_dt,
            "event": event_label,
            "type": "Acquisition",
            "details": f"Duration: {(end_dt - start_dt).total_seconds() / 3600:.1f} hours",
            "data": acquisition,
        }
    except Exception as e:
        print(f"[parse_acquisition] Error parsing acquisition: {e}")
        return None


def parse_perfusion(proc, surgery_date=None):
    """Parse perfusion procedure (surgery sub-procedure)."""
    if not proc:
        return None

    # Perfusion procedures are typically part of a surgery
    # If surgery_date is provided, use it; otherwise look for start_date
    start_date = surgery_date or proc.get("start_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)
        # Perfusion is typically quick (1-2 hours)
        end_dt = start_dt + timedelta(hours=1)

        protocol_id = proc.get("protocol_id", "Unknown")
        specimen_ids = proc.get("output_specimen_ids", [])
        specimen_str = ", ".join(specimen_ids) if specimen_ids else "Unknown"

        return {
            "start": start_dt,
            "end": end_dt,
            "event": "Perfusion",
            "type": "Perfusion",
            "details": f"Perfusion (specimen: {specimen_str})",
            "data": proc,
        }
    except Exception as e:
        print(f"[parse_perfusion] Error parsing perfusion: {e}")
        return None


def parse_brain_injection(proc, surgery_date=None):
    """Parse brain injection procedure (surgery sub-procedure)."""
    if not proc:
        return None

    # Brain injection procedures are typically part of a surgery
    start_date = surgery_date or proc.get("start_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)
        # Brain injection is typically 1-3 hours
        end_dt = start_dt + timedelta(hours=2)

        # Extract injection materials info
        materials = proc.get("injection_materials", [])
        material_names = []
        for mat in materials:
            if mat and mat.get("object_type") == "Viral material":
                name = mat.get("name", "Unknown")
                material_names.append(name)

        materials_str = ", ".join(material_names) if material_names else "Unknown"

        # Extract position info
        positions = proc.get("relative_position", [])
        position_str = ", ".join(positions) if positions else "Unknown"

        return {
            "start": start_dt,
            "end": end_dt,
            "event": "Brain injection",
            "type": "Brain injection",
            "details": f"Brain injection ({position_str}): {materials_str}",
            "data": proc,
        }
    except Exception as e:
        print(f"[parse_brain_injection] Error parsing brain injection: {e}")
        return None


def parse_generic_surgery_procedure(proc, surgery_date=None):
    """Parse generic surgery procedure (surgery sub-procedure)."""
    if not proc:
        return None

    # Generic surgery procedures are part of a surgery
    start_date = surgery_date or proc.get("start_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)
        # Generic procedure duration varies
        end_dt = start_dt + timedelta(hours=2)

        description = proc.get("description", "Generic surgery procedure")
        notes = proc.get("notes", "")

        details_str = description
        if notes:
            details_str += f" - {notes}"

        return {
            "start": start_dt,
            "end": end_dt,
            "event": "Generic surgery procedure",
            "type": "Generic surgery procedure",
            "details": details_str,
            "data": proc,
        }
    except Exception as e:
        print(f"[parse_generic_surgery_procedure] Error parsing generic surgery procedure: {e}")
        return None


def parse_specimen_procedure(proc):
    """Parse specimen procedure."""
    if not proc:
        return None

    start_date = proc.get("start_date")
    end_date = proc.get("end_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)

        # Use end_date if available, otherwise use 1 day duration
        if end_date:
            end_dt = normalize_timestamp(end_date)
        else:
            end_dt = start_dt + NO_END_DATE_DURATION

        procedure_type = proc.get("procedure_type", "Unknown")
        procedure_name = proc.get("procedure_name", "Unknown")
        specimen_id = proc.get("specimen_id", "Unknown")

        # Extract reagent info if available
        details_list = proc.get("procedure_details", [])
        reagents = []
        for detail in details_list:
            if detail and detail.get("object_type") == "Reagent":
                reagent_name = detail.get("name", "Unknown")
                lot_number = detail.get("lot_number")
                if lot_number:
                    reagents.append(f"{reagent_name} (lot: {lot_number})")
                else:
                    reagents.append(reagent_name)

        # Build details string
        details_parts = [procedure_name]
        if reagents:
            details_parts.append(f"Reagents: {', '.join(reagents)}")
        if specimen_id != "Unknown":
            details_parts.append(f"Specimen: {specimen_id}")

        details_str = " - ".join(details_parts)

        return {
            "start": start_dt,
            "end": end_dt,
            "event": procedure_type,
            "type": procedure_type,
            "details": details_str,
            "data": proc,
        }
    except Exception as e:
        print(f"[parse_specimen_procedure] Error parsing specimen procedure: {e}")
        return None


def parse_fiber_implant(proc, surgery_date=None):
    """Parse fiber implant procedure (surgery sub-procedure)."""
    if not proc:
        return None

    # Fiber implant procedures are typically part of a surgery
    start_date = surgery_date or proc.get("start_date")

    if not start_date:
        return None

    try:
        start_dt = normalize_timestamp(start_date)
        # Fiber implant surgery is typically 2-3 hours
        end_dt = start_dt + timedelta(hours=2)

        # Check for device_config (fiber probe info)
        device_config = proc.get("device_config")
        if device_config:
            device_name = device_config.get("device_name", "Unknown")
            details_str = f"Fiber probe implant ({device_name})"
        else:
            details_str = "Probe implant"

        return {
            "start": start_dt,
            "end": end_dt,
            "event": "Probe implant",
            "type": "Probe implant",
            "details": details_str,
            "data": proc,
            "has_fiber_visualization": True,  # Flag to indicate special visualization available
        }
    except Exception as e:
        print(f"[parse_fiber_implant] Error parsing fiber implant: {e}")
        return None

File: web/src/subject/parsers.js
/**
 * subject-parsers.js — Event parsers for the subject timeline.
 *
 * Ported from:
 *   src/zombie/subject_contents/procedures/parsers.py
 *   src/zombie/subject_contents/procedures/brain_injection_parser.py
 *   src/zombie/subject_contents/procedures/fiber_implant_parser.py
 *
 * All functions are pure and side-effect-free; suitable for unit testing
 * in a Node environment without DOM.
 */

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Default event duration (ms) when no end date is provided — 1 day. */
export const NO_END_DATE_DURATION_MS = 24 * 60 * 60 * 1000;

/**
 * Color map for timeline event types (matches CSS class usage in timeline).
 * Used by subject-timeline.js and subject-details.js.
 */
export const EVENT_COLORS = {
  Birth: '#9C27B0',
  Surgery: '#F44336',
  Acquisition: '#2196F3',
  Session: '#4CAF50',
  Perfusion: '#FF5722',
  'Brain injection': '#E91E63',
  'Probe implant': '#00BCD4',
  'Generic surgery procedure': '#FF9800',
  Fixation: '#607D8B',
  Delipidation: '#795548',
  'Refractive index matching': '#009688',
};

// ---------------------------------------------------------------------------
// Shared utilities
// ---------------------------------------------------------------------------

/**
 * Safely convert a value to a finite float.
 *
 * @param {unknown} value
 * @param {number} [defaultVal=0]
 * @returns {number}
 */
export function safeFloat(value, defaultVal = 0) {
  if (value == null) return defaultVal;
  const n = Number(value);
  return isFinite(n) ? n : defaultVal;
}

/**
 * Normalise a timestamp string or Date to a JS Date.
 *
 * @param {string|Date|null|undefined} ts
 * @returns {Date|null}
 */
export function normalizeTimestamp(ts) {
  if (!ts) return null;
  const d = ts instanceof Date ? ts : new Date(ts);
  return isNaN(d.getTime()) ? null : d;
}

// ---------------------------------------------------------------------------
// Timeline event parsers
// ---------------------------------------------------------------------------

/**
 * Parse birth event from a subject record.
 *
 * @param {object} subject - Subject portion of a DocDB record.
 * @returns {object|null} Timeline event object or null.
 */
export function parseBirth(subject) {
  if (!subject) return null;
  const details = subject.subject_details ?? {};
  const dob = details.date_of_birth;
  if (!dob) return null;

  const start = normalizeTimestamp(dob);
  if (!start) return null;

  return {
    start,
    end: new Date(start.getTime() + NO_END_DATE_DURATION_MS),
    event: 'Birth',
    type: 'Birth',
    details: `Subject ${subject.subject_id ?? 'Unknown'} born`,
    data: details,
    dateOnly: true,
  };
}

/**
 * Parse a Surgery/Procedure event from a subject_procedures entry.
 *
 * @param {object} proc
 * @returns {object|null}
 */
export function parseProcedure(proc) {
  if (!proc) return null;
  const procType = proc.object_type ?? 'Procedure';
  const start = normalizeTimestamp(proc.start_date);
  if (!start) return null;

  const end = proc.end_date
    ? (normalizeTimestamp(proc.end_date) ?? new Date(start.getTime() + NO_END_DATE_DURATION_MS))
    : new Date(start.getTime() + NO_END_DATE_DURATION_MS);

  let detailStr;
  if (procType === 'Surgery') {
    const subProcs = (proc.procedures ?? []).filter(Boolean);
    const parts = subProcs.map((sp) => sp.object_type ?? 'Unknown');
    detailStr = parts.length ? parts.join(', ') : 'Surgery';
  } else {
    detailStr = procType;
  }

  return { start, end, event: procType, type: procType, details: detailStr, data: proc, dateOnly: true };
}

/**
 * Parse an acquisition event (with start + end times).
 *
 * @param {object} acquisition
 * @returns {object|null}
 */
export function parseAcquisition(acquisition) {
  if (!acquisition) return null;
  const start = normalizeTimestamp(acquisition.acquisition_start_time);
  const end = normalizeTimestamp(acquisition.acquisition_end_time);
  if (!start || !end) return null;

  const acqType = acquisition.acquisition_type ?? acquisition.session_type ?? 'Acquisition';
  const protocol = acquisition.protocol_name ?? '';
  const label = protocol ? `${acqType} (${protocol})` : acqType;
  const durationHrs = (end - start) / 3_600_000;

  return {
    start,
    end,
    event: label,
    type: 'Acquisition',
    details: `Duration: ${durationHrs.toFixed(1)} hours`,
    data: acquisition,
  };
}

/**
 * Parse a session from a data_description record.
 *
 * @param {object} dataDesc
 * @returns {object|null}
 */
export function parseSession(dataDesc) {
  if (!dataDesc) return null;
  const start = normalizeTimestamp(dataDesc.creation_time);
  if (!start) return null;

  return {
    start,
    end: new Date(start.getTime() + 4 * 3_600_000),
    event: 'Session',
    type: 'Session',
    details: `Data session: ${dataDesc.name ?? 'Unknown'}`,
    data: dataDesc,
  };
}

/**
 * Parse a perfusion sub-procedure.
 *
 * @param {object} proc
 * @param {string|Date|null} [surgeryDate]
 * @returns {object|null}
 */
export function parsePerfusion(proc, surgeryDate = null) {
  if (!proc) return null;
  const start = normalizeTimestamp(surgeryDate ?? proc.start_date);
  if (!start) return null;

  const specimens = proc.output_specimen_ids ?? [];
  return {
    start,
    end: new Date(start.getTime() + 3_600_000),
    event: 'Perfusion',
    type: 'Perfusion',
    details: `Perfusion (specimen: ${specimens.length ? specimens.join(', ') : 'Unknown'})`,
    data: proc,
    dateOnly: true,
  };
}

/**
 * Parse a brain injection sub-procedure.
 *
 * @param {object} proc
 * @param {string|Date|null} [surgeryDate]
 * @returns {object|null}
 */
export function parseBrainInjection(proc, surgeryDate = null) {
  if (!proc) return null;
  const start = normalizeTimestamp(surgeryDate ?? proc.start_date);
  if (!start) return null;

  const materials = (proc.injection_materials ?? [])
    .filter((m) => m?.object_type === 'Viral material')
    .map((m) => m.name ?? 'Unknown');
  const positions = proc.relative_position ?? [];

  return {
    start,
    end: new Date(start.getTime() + 2 * 3_600_000),
    event: 'Brain injection',
    type: 'Brain injection',
    details: `Brain injection (${positions.join(', ') || 'Unknown'}): ${materials.join(', ') || 'Unknown'}`,
    data: proc,
    dateOnly: true,
  };
}

/**
 * Parse a generic surgery sub-procedure.
 *
 * @param {object} proc
 * @param {string|Date|null} [surgeryDate]
 * @returns {object|null}
 */
export function parseGenericSurgeryProcedure(proc, surgeryDate = null) {
  if (!proc) return null;
  const start = normalizeTimestamp(surgeryDate ?? proc.start_date);
  if (!start) return null;

  const description = proc.description ?? 'Generic surgery procedure';
  const notes = proc.notes ?? '';

  return {
    start,
    end: new Date(start.getTime() + 2 * 3_600_000),
    event: 'Generic surgery procedure',
    type: 'Generic surgery procedure',
    details: notes ? `${description} - ${notes}` : description,
    data: proc,
    dateOnly: true,
  };
}

/**
 * Parse a specimen procedure.
 *
 * @param {object} proc
 * @returns {object|null}
 */
export function parseSpecimenProcedure(proc) {
  if (!proc) return null;
  const start = normalizeTimestamp(proc.start_date);
  if (!start) return null;

  const end = proc.end_date
    ? (normalizeTimestamp(proc.end_date) ?? new Date(start.getTime() + NO_END_DATE_DURATION_MS))
    : new Date(start.getTime() + NO_END_DATE_DURATION_MS);

  const procedureType = proc.procedure_type ?? 'Unknown';
  const procedureName = proc.procedure_name ?? 'Unknown';
  const specimenId = proc.specimen_id ?? 'Unknown';
  const reagents = (proc.procedure_details ?? [])
    .filter((d) => d?.object_type === 'Reagent')
    .map((d) => (d.lot_number ? `${d.name ?? 'Unknown'} (lot: ${d.lot_number})` : (d.name ?? 'Unknown')));

  const parts = [procedureName];
  if (reagents.length) parts.push(`Reagents: ${reagents.join(', ')}`);
  if (specimenId !== 'Unknown') parts.push(`Specimen: ${specimenId}`);

  return { start, end, event: procedureType, type: procedureType, details: parts.join(' - '), data: proc, dateOnly: true };
}

/**
 * Parse a fiber/probe implant sub-procedure.
 *
 * @param {object} proc
 * @param {string|Date|null} [surgeryDate]
 * @returns {object|null}
 */
export function parseFiberImplant(proc, surgeryDate = null) {
  if (!proc) return null;
  const start = normalizeTimestamp(surgeryDate ?? proc.start_date);
  if (!start) return null;

  const dc = proc.device_config;
  const detailStr = dc ? `Fiber probe implant (${dc.device_name ?? 'Unknown'})` : 'Probe implant';

  return {
    start,
    end: new Date(start.getTime() + 2 * 3_600_000),
    event: 'Probe implant',
    type: 'Probe implant',
    details: detailStr,
    data: proc,
    hasFiberVisualization: true,
    dateOnly: true,
  };
}

/**
 * Build the full list of timeline events from a subject data bundle.
 *
 * @param {object} subjectData - Bundle: { subject, procedures, acquisitions }
 * @returns {Array<object>} Sorted array of timeline event objects.
 */
export function buildTimelineEvents(subjectData) {
  if (!subjectData) return [];
  const events = [];

  // Birth
  const birth = parseBirth(subjectData.subject ?? {});
  if (birth) events.push(birth);

  // Subject procedures (top-level: Surgery, Headframe, etc.)
  const subjectProcedures = subjectData.procedures?.subject_procedures ?? [];
  for (const proc of subjectProcedures) {
    const ev = parseProcedure(proc);
    if (ev) events.push(ev);
  }

  // Specimen procedures
  const specimenProcedures = subjectData.procedures?.specimen_procedures ?? [];
  for (const proc of specimenProcedures) {
    const ev = parseSpecimenProcedure(proc);
    if (ev) events.push(ev);
  }

  // Acquisitions
  const acquisitions = subjectData.acquisitions ?? [];
  for (const acq of acquisitions) {
    const ev = parseAcquisition(acq);
    if (ev) events.push(ev);
  }

  events.sort((a, b) => a.start - b.start);
  return events;
}

// ---------------------------------------------------------------------------
// Brain injection helpers (ported from brain_injection_parser.py)
// ---------------------------------------------------------------------------

/**
 * Extract AP/ML/DV coordinates from a brain injection procedure.
 *
 * @param {object} injectionProc
 * @returns {[number, number, number]} [ap, ml, dv]
 */
export function extractInjectionCoordinates(injectionProc) {
  const coordinates = injectionProc?.coordinates ?? [];
  if (!coordinates.length) return [0, 0, 0];
  const coordSet = coordinates[0];
  if (!Array.isArray(coordSet) || !coordSet.length) return [0, 0, 0];
  const translation = coordSet[0];
  if (!translation || translation.object_type !== 'Translation') return [0, 0, 0];
  const vals = translation.translation ?? [];
  return vals.length >= 3
    ? [safeFloat(vals[0]), safeFloat(vals[1]), safeFloat(vals[2])]
    : [0, 0, 0];
}

/**
 * Extract viral material list from an injection procedure.
 *
 * @param {object} injectionProc
 * @returns {Array<object>}
 */
export function extractInjectionMaterials(injectionProc) {
  return (injectionProc?.injection_materials ?? [])
    .filter((m) => m?.object_type === 'Viral material')
    .map((m) => {
      const tars = m.tars_identifiers ?? {};
      return {
        name: m.name ?? 'Unknown',
        titer: m.titer ?? 'Unknown',
        titerUnit: m.titer_unit ?? '',
        tarsId: tars.virus_tars_id ?? '',
        lotNumber: tars.prep_lot_number ?? '',
      };
    });
}

/**
 * Extract injection dynamics from a procedure.
 *
 * @param {object} injectionProc
 * @returns {object|null}
 */
export function extractInjectionDynamics(injectionProc) {
  const dynamics = injectionProc?.dynamics ?? [];
  if (!dynamics.length) return null;
  const dyn = dynamics[0];
  if (!dyn) return null;
  return {
    profile: dyn.profile ?? 'Unknown',
    volume: safeFloat(dyn.volume),
    volumeUnit: dyn.volume_unit ?? 'nL',
    duration: dyn.duration != null ? safeFloat(dyn.duration) : null,
    durationUnit: dyn.duration_unit ?? 's',
  };
}

/**
 * Get numeric injection index from name for sort ordering.
 *
 * @param {object} injectionData
 * @returns {number}
 */
export function getInjectionIndex(injectionData) {
  const name = injectionData?.name ?? '';
  if (name.includes('_')) {
    const n = parseInt(name.split('_').at(-1), 10);
    if (!isNaN(n)) return n;
  }
  return 999;
}

/**
 * Extract all brain injection data from a surgery.
 *
 * @param {object} surgeryData
 * @returns {Array<object>}
 */
export function extractInjectionsFromSurgery(surgeryData) {
  return (surgeryData?.procedures ?? [])
    .map((proc, idx) => {
      if (!proc || proc.object_type !== 'Brain injection') return null;
      const [ap, ml, dv] = extractInjectionCoordinates(proc);
      const materials = extractInjectionMaterials(proc);
      const dynamics = extractInjectionDynamics(proc);
      return {
        name: `Injection_${idx}`,
        ap,
        ml,
        dv,
        unit: 'millimeter',
        reference: proc.coordinate_system_name ?? 'Unknown',
        position: (proc.relative_position ?? []).join(', ') || 'Unknown',
        materials,
        materialNames: materials.map((m) => m.name),
        dynamics,
        protocolId: proc.protocol_id ?? 'Not specified',
      };
    })
    .filter(Boolean);
}

/**
 * Check if a surgery has any brain injection procedures.
 *
 * @param {object} surgeryData
 * @returns {boolean}
 */
export function hasBrainInjections(surgeryData) {
  return (surgeryData?.procedures ?? []).some((p) => p?.object_type === 'Brain injection');
}

// ---------------------------------------------------------------------------
// Fiber implant helpers (ported from fiber_implant_parser.py)
// ---------------------------------------------------------------------------

/**
 * Extract fiber metadata from a device config.
 *
 * @param {object} deviceConfig
 * @returns {object}
 */
export function extractFiberMetadata(deviceConfig) {
  let ap = 0, ml = 0, dv = null, depth = null, angle = 0;
  for (const t of deviceConfig?.transform ?? []) {
    if (t?.object_type === 'Translation') {
      const vals = t.translation ?? [];
      if (vals.length >= 3) { ap = safeFloat(vals[0]); ml = safeFloat(vals[1]); dv = safeFloat(vals[2]); }
      if (vals.length >= 4) { depth = Math.abs(safeFloat(vals[3])); }
    } else if (t?.object_type === 'Rotation') {
      const r = t.rotation ?? [];
      if (r.length >= 1) angle = safeFloat(r[0]);
    }
  }
  return {
    name: deviceConfig?.device_name ?? 'Unknown',
    ap,
    ml,
    dv,
    depth,
    angle,
    unit: 'millimeter',
    reference: (deviceConfig?.coordinate_system ?? {}).origin ?? 'Bregma',
    targetedStructure: (deviceConfig?.primary_targeted_structure ?? {}).name
      ?? 'Not specified in surgical request form',
  };
}

/**
 * Extract all fiber implant data from a surgery.
 *
 * @param {object} surgeryData
 * @returns {Array<object>}
 */
export function extractFibersFromSurgery(surgeryData) {
  return (surgeryData?.procedures ?? [])
    .filter((p) => p?.object_type === 'Probe implant' && p.device_config)
    .map((p) => extractFiberMetadata(p.device_config));
}

/**
 * Check if a surgery has any fiber/probe implant procedures.
 *
 * @param {object} surgeryData
 * @returns {boolean}
 */
export function hasFiberImplants(surgeryData) {
  return (surgeryData?.procedures ?? []).some((p) => p?.object_type === 'Probe implant');
}

/**
 * Get numeric fiber index from name for sort ordering.
 *
 * @param {object} fiber
 * @returns {number}
 */
export function getFiberIndex(fiber) {
  const n = parseInt((fiber?.name ?? '').split('_').at(-1), 10);
  return isNaN(n) ? 999 : n;
}
